lib/spack/spack/mirror.py

```diff
@@ -98,6 +98,63 @@ def from_dict(d, name=None):
         else:
             return Mirror(d['fetch'], d['push'], name)
 
+    @staticmethod
+    def from_args(args):
+        result = None
+
+        if ((args.directory and args.mirror_name) or
+                (args.mirror_name and args.mirror_url) or
+                (args.mirror_url and args.directory)):
+            raise ValueError('no more than one of "directory",'
+                             ' "mirror_url", and "mirror_name" allowed')
+
+        if args.directory:
+            location = args.directory
+
+            # User meant to provide a path to a local directory.
+            # Ensure that they did not accidentally pass a URL.
+            scheme = url_util.parse(location, scheme='<missing>').scheme
+            if scheme != '<missing>':
+                raise ValueError(
+                    '"directory" must be a local path (got a URL)')
+
+            # User meant to provide a path to a local directory.
+            # Ensure that the mirror lookup does not mistake it for a named
+            # mirror.
+            location = 'file://' + location
+            result = spack.mirror.MirrorCollection().lookup(location)
```

Comment:
This was hard for me to understand and I had to read code carefully to get what's going on. A few suggestions below.

---

I think people would expect the default constructor of a mapping to start empty, like the built-in `dict()`. Here:
```python
spack.mirror.MirrorCollection()
```
reads instead from the global configuration and creates an object with items in it. Besides being unexpected, this default logic couples the class to a global state. Can we extract that logic into a separate function:
```python
def mirror_collection_from_config():
    ...
```
and have the default constructor behave as expected?

---

The `MirrorCollection.lookup` method acts like a factory for mirror items. This is totally unexpected too:
```python
Python 3.8.10, Linux x86_64
>>> import spack.mirror
>>> c = spack.mirror.MirrorCollection()
>>> s = c.lookup('/tmp/mirror')
>>> s
Mirror(fetch_url='/tmp/mirror')
>>> s in c
False
```
Mappings have already a `lookup` method, which is "get" and accepts a default. I think `MirrorCollection.lookup` should be removed in favor of `MirrorCollection.get`.

---

Finally a note that if the class is not meant to be immutable, it should be better to derive `MirrorCollection` from `MutableMapping`.

Reply:
These are very valid considerations.  I will follow up with some changes.

Reply:
Some points I'm reminded of, having taken a second look at this:

`MirrorCollection` already has a `get()` method that works the same as with other Mappings.  `MirrorCollection.lookup()` works like `get()`, except it has a default return value that depends on the argument passed.  A more verbose name might be `get_or_return_anonymous()`.

Also, `MirrorCollections` *are* immutable.  That's why there is no `__setitem__()` defined.  They're meant to represent a preexisting collection of mirrors that come from a Spack configuration.  This is also why the default constructor doesn't return an empty collection; that would make no sense.  I think the naming should be improved to make it clearer that this class isn't meant to be used as a regular container type.

Perhaps `MirrorConfiguration` or `MirrorGroup` would be a better name?

Reply:
> Also, MirrorCollections are immutable. 

That is fine, but they should be initialized from an iterable of some sort (like `frozenset` for instance).

> MirrorCollection already has a get() method that works the same as with other Mappings.

Can't we delete `lookup` and use:
```python
c = spack.mirror.MirrorCollection(iterable)
s = c.get('/tmp/mirror', Mirror(fetch_url='/tmp/mirror'))
```
?

Reply:
Yes, I see.  I will look into making those changes.